File: leetcode/Medium/279.perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [i for i in range(n+1)]
        for i in range(4, n+1):
            x = 1
            while x * x <= i:
                tmp = x * x
                if tmp > i:
                    break
                else:
                    dp[i] = min(dp[i], 1 + dp[i - tmp])
                x += 1

        return dp[n]

File: leetcode/Medium/279.perfect-squares.cpp
class Solution {
 public:
  int numSquares(int n) {
    int* dp = new int[n + 1];
    dp[0] = 0;
    dp[1] = 1;

    for (int i = 2; i <= n; i++) {
      dp[i] = i;
      for (int x = 1; x * x <= i; x++) {
        int tmp = x * x;
        if (tmp > i) {
          break;
        } else {
          dp[i] = min(dp[i], 1 + dp[i - tmp]);
        }
      }
    }

    int res = dp[n];
    delete[] dp;

    return res;
  }
};
